sdk/python/feast/pyspark/launchers/k8s/k8s.py

```diff
@@ -0,0 +1,344 @@
+
+        Raises:
+            SparkJobFailure: The spark job submission failed, encountered error
+                during execution, or timeout.
+
+        Returns:
+            RetrievalJob: wrapper around remote job that returns file uri to the result file.
+        """
+
+        with open(job_params.get_main_file_path()) as f:
+            pyspark_script = f.read()
+
+        pyspark_script_path = urlunparse(
+            self._get_staging_client().upload_fileobj(
+                BytesIO(pyspark_script.encode("utf8")),
+                local_path="historical_retrieval.py",
+                remote_path_prefix=self._staging_location,
+                remote_path_suffix=".py",
+            )
+        )
+
+        job_id = _generate_job_id()
+
+        resource = _prepare_job_resource(
+            job_template=self._resource_template,
+            job_id=job_id,
+            job_type=HISTORICAL_RETRIEVAL_JOB_TYPE,
+            main_application_file=pyspark_script_path,
+            main_class=None,
+            packages=[],
+            jars=[],
+            extra_metadata={METADATA_OUTPUT_URI: job_params.get_destination_path()},
+            arguments=job_params.get_arguments(),
+        )
+
+        job_info = _submit_job(
+            api=self._api, resource=resource, namespace=self._namespace,
+        )
+
+        return cast(RetrievalJob, self._job_from_job_info(job_info))
+
+    def _upload_jar(self, jar_path: str) -> str:
+        if jar_path.startswith("s3://") or jar_path.startswith("s3a://"):
+            return jar_path
+        elif jar_path.startswith("file://"):
+            local_jar_path = urlparse(jar_path).path
+        else:
+            local_jar_path = jar_path
+        with open(local_jar_path, "rb") as f:
+            return urlunparse(
+                self._get_staging_client().upload_fileobj(
+                    f,
+                    local_jar_path,
+                    remote_path_prefix=self._staging_location,
+                    remote_path_suffix=".jar",
+                )
+            )
+
+    def offline_to_online_ingestion(
+        self, ingestion_job_params: BatchIngestionJobParameters
+    ) -> BatchIngestionJob:
+        """
+        Submits a batch ingestion job to a Spark cluster.
+
+        Raises:
+            SparkJobFailure: The spark job submission failed, encountered error
+                during execution, or timeout.
+
+        Returns:
+            BatchIngestionJob: wrapper around remote job that can be used to check when job completed.
+        """
+
+        jar_s3_path = self._upload_jar(ingestion_job_params.get_main_file_path())
+
+        job_id = _generate_job_id()
+
+        resource = _prepare_job_resource(
+            job_template=self._resource_template,
+            job_id=job_id,
+            job_type=OFFLINE_TO_ONLINE_JOB_TYPE,
+            main_application_file=jar_s3_path,
+            main_class="feast.ingestion.IngestionJob",
```

Comment:
shouldn't main_class be retrieved from `job_params`?

Reply:
fixed